# Maximizing Ad Revenue with the Free-App-Only Approach to the App Store and Google Play Markets

In this project, I'm playing the role of a data analyst at a company that creates free applications for the App Store and Google Play markets. Since our apps are free, we can only make money from in-app ads, meaning we will disregard all paid app data.

The best way to accrue ad revenue is for as many people to see our ads as possible. Therefore, our goal is to analyze the apps in this market to better help our company figure out what types of free apps attract the most users in order to increase our profit using the free-app-only approach.

## 1. Opening the Datasets

We need to open the two datasets in Jupyter Notebooks to start our analysis. First, we open the datasets using Python's built-in *open()* function. 

In [1]:
apple_file = open("AppleStore.csv")
google_file = open("googleplaystore.csv")

Next, we import the *reader* function from the *csv* library in order to read the *.csv* datasets we just opened.

In [2]:
from csv import reader
read_apple_file = reader(apple_file)
read_google_file = reader(google_file)

Finally, we turn the datasets into list-of-lists using the built-in *list()* function.

In [3]:
app_data = list(read_apple_file)
goog_data = list(read_google_file)

# For the rest of this project, app_data and 
# goog_data are the App Store and Google Play 
# datasets respectively.

## 2. Exploring the Datasets

Our company has given us this function to help us get started. We can use *explore_data()* to specify which chunk of our dataset we would like to print out and view, as well as an option to count the number of rows and columns if we want.

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
   
    # Here we use a for loop to print each list of
    # the list-of-lists (each row of the dataset).

    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    # And this if statement will calculate the number
    # of rows and columns in the dataset if we set
    # the boolean rows_and_columns to True.
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

Now, by calling the function, we can easily display the first few rows (and skip the header row) by passing in start = *1* and end = *4*. We can also change the boolean variable *rows_and_columns* to *True* in order to see the number of rows and columns in the datasets.

In [5]:
explore_data(app_data,1,4,True) # Choose 1 to exclude the header row.
print('\n')
explore_data(goog_data,1,4,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7198
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1,

It is also useful for us to know the column names so we can identify which information is the most important for our approach. We'll do this by printing the header rows.

In [6]:
print(app_data[0], "\n")
print(goog_data[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


**To save the trouble of defining all 29 column names, links to the [Apple](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) and [Google](https://www.kaggle.com/lava18/google-play-store-apps) dataset documentations have been provided.**

From the documentation provided, and from our current interest of making free apps that appeal to the widest audience, we're interested in *track_name, currency, price, rating_count_tot, rating_count_ver, cont_rating, and prime_genre* from the Apple dataset.

Following this logic, we would be interested in *Category, Reviews, Installs, Type, Price, Content Rating, and Genres* from the Google dataset. We're not interested in positive 
review data since our goal is to simply reach the most people we can.

## 3. Fixing Errors

Before we begin, we found out on a [discussion board](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) that there's an error in the Google dataset around either line *10472* or *10473*. Let's investigate.

In [7]:
print(goog_data[10472], "\n")
print(goog_data[10473], "\n")
print(goog_data[1], "\n")

# By using the built in length function, we can count
# how many items are in a list and compare them. I also
# chose to include line 1 as a control in this experiment.

print(len(goog_data[10472]), "\n")
print(len(goog_data[10473]), "\n")
print(len(goog_data[1]))

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up'] 

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

13 

12 

13


We can see that the error was indeed at line *10473* because there are only 12 items in the list instead of 13. It's best not to include this row of the dataset since it's missing two values from a column that we're interested in, so let's delete it.

In [8]:
print(goog_data[10474])
del goog_data[10473]
# DO NOT RUN THIS MORE THAN ONCE!!

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


In [9]:
print(goog_data[10473])

# Upon printing, we should see the old row 10474
# that we printed above has now taken the place 
# of our deleted row 10473.

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


## 4. Removing Duplicate Entries

Looking further through the discussion board over the Google dataset, there's a [post](https://www.kaggle.com/lava18/google-play-store-apps/discussion/120506) about duplicate entries. Let's test a random  app name with a *for* loop and see if we come up with any duplicates.

In [10]:
for row in goog_data[1:]: # index 0 contains the header row.
    app_name = row[0] # index 0 contains the app name.
    if app_name == 'Clash of Clans':
        print(row)

['Clash of Clans', 'GAME', '4.6', '44891723', '98M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Strategy', 'July 15, 2018', '10.322.16', '4.1 and up']
['Clash of Clans', 'GAME', '4.6', '44891723', '98M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Strategy', 'July 15, 2018', '10.322.16', '4.1 and up']
['Clash of Clans', 'GAME', '4.6', '44893888', '98M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Strategy', 'July 15, 2018', '10.322.16', '4.1 and up']
['Clash of Clans', 'FAMILY', '4.6', '44881447', '98M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Strategy', 'July 15, 2018', '10.322.16', '4.1 and up']


Since the *for* loop returned 4 results, we know that there are 3 duplicate entries of *Clash of Clans*. Knowing this, let's edit the *for* loop so we can count all of the instances of duplicate entries.

In [11]:
duplicate_counter = 0 # Initializing the duplicate counter at zero.
unique_entries = [] # Empty list that will store the proper entries.
for row in goog_data[1:]:
    if row[0] in unique_entries:
        duplicate_counter += 1 # If our row is alreay there, it's a copy.
    else:
        unique_entries.append(row[0]) # If it isn't, add it to the list.
        
print(duplicate_counter) # This will be an integer since we used a counter.
print(len(unique_entries)) # The length of this list is a comparable integer.

1181
9659


With this output, we can see there's an astounding 1,181 duplicate entries in our dataset. There are also 9,659 unique entries, but we can't just take this for face value. Let's make sure these numbers make sense.

In [12]:
theoretical_length = duplicate_counter + len(unique_entries)
print(theoretical_length)

actual_length = len(goog_data[1:]) # Exclude the header row.
print(actual_length)

10840
10840


Now that we know our numbers are correct, we can get rid of these duplicate entries. We don't want to randomly delete entries since we could be losing valuable data. Instead, we'll delete duplicates that have **less reviews**, since logically the number of reviews a popular app has shouldn't decrease over time. *People can't rescind a review like they can rescind a rating.*

We first need to implement our strategy by creating an empty dictionary that stores only unique entries. However, we'll need to update this dictionary with a *for* loop whenever we find a duplicate entry that has a higher number of reviews than a unique entry.

In [13]:
reviews_max = {} # Create an empty dictionary.
for row in goog_data[1:]:
    name = row[0]
    n_reviews = float(row[3]) # Index 3 is Reviews. We can't act
                              # on a string, so we conver to float.
    
    # This first if statement will only update the dictionary
    # when we find a duplicate entry with more reviews than our
    # current entry.
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    # This second if statement will create an entry in the dictionary
    # if and only if we find a unique entry.
    
    if name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max)) # The length of our dictionary should be 9659.

9659


Our output is *9659*, which confirms that our technique worked in creating a dictionary with no duplicate entries. Now, we need to actually delete these entries from the dataset.

In [14]:
android_clean = [] # Two empty lists to help with the deletion process.
already_added = []

for row in goog_data[1:]: # This code block is unchanged from before.
    name = row[0]
    n_reviews = float(row[3])
    
    # But now we can use our dictionary to help us. We'll first check
    # if the number of reviews in a certain row of data is equal to
    # that of a unique value in our dictionary. Then, we'll see if it
    # has already been added to our list defined above. If not, the
    # row will be added to our android_clean list.
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)        

print(len(android_clean)) # Again, we should get 9659.
print(len(already_added)) # And this should also be 9659.

9659
9659


The reason *android_clean* and *already_added* are the same length is because both lists are only appended when the *if* statement is fulfilled. Otherwise, duplicate entries are ignored. 

Notice the difference between both lists: *android_clean* contains all unique rows (with the max number of reviews) from the Google Play dataset. But *already_added* just contains the name of apps that have already been added to the clean list.

In [15]:
print(android_clean[0],"\n")
print(already_added[0])

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

Photo Editor & Candy Camera & Grid & ScrapBook


Before we move on, it's necessary to check if there's any discussion of incorrect data or duplicate entries on the Apple dataset discussions board. It seems that upon [looking there](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion/90409), in a rare instance, the dataset is completely clean and free of duplicates and errors.

## 5. Removing Non-English Apps

Our company is interested in almost all aspects of marketing free-apps, but we want to keep our focus on an English-speaking audience. Therefore, we need to remove all non-English entries in the new *android_clean* dataset so that these apps won't interfere with our analysis. 

We know that Python matches an integer from *0* to *127* with the letters of the English alphabet, so let's make a function that searches through the string of an app name, checks each letter and it's corresponding integer match, and reports if any letter is above *127* in value.

To check the integer corresponding to a character in a string, we'll use Pythons built-in *ord()* function (the integer is known as a **Unicode** value).

In [16]:
def english_or_not(name_of_app):
    
    for char in name_of_app:
        if ord(char) > 127:
            return False # If integer > 127, return false.
    
    return True # If 0 ≤ integer ≤ 127, return true.
                # This means the if statement was ignored.

# Now, we can test this function by calling it and passing
# through a few app names our company provided us to check.

print(english_or_not('Instagram'))
print(english_or_not('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_or_not('Docs To Go™ Free Office Suite'))
print(english_or_not('Instachat 😜'))

True
False
False
False


It looks like our function worked perfectly and *Instagram* is considered an English-speaking app while *爱奇艺PPS -《欢乐颂2》电视剧热播* is not. However, we have a problem in that *Docs To Go™ Free Office Suite* and *Instachat 😜* are considered non-English-speaking when they clearly **are** English-speaking.

In [17]:
print(ord('播'))
print(ord('™'))
print(ord('😜'))

25773
8482
128540


The problem is that just like Chinese characters, emoji are considered non-English characters. But since plenty of English-speaking apps use emoji, we'll need to relax the requirements of our function to return *False* so that we can allow English-speaking apps with emoji in the title to stay in our dataset. We'll use a counter to limit the number of emoji to *3*, any more and we'll return *False*.

In [18]:
def english_or_not(name_of_app):
    emoji_counter = 0 # 
    for char in name_of_app:
        if ord(char) > 127:
            emoji_counter += 1 # If integer > 127, iterate counter.
    
    if emoji_counter > 3: # Only returns true if counter > 3.
        return False
    
    return True # Only returns true if counter ≤ 3.

print(english_or_not('Instagram'))
print(english_or_not('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_or_not('Docs To Go™ Free Office Suite'))
print(english_or_not('Instachat 😜'))

True
False
True
True


*爱奇艺PPS -《欢乐颂2》电视剧热播* is still considered a non-English app since there are plenty more than 3 special characters in the title. But now, *Docs To Go™ Free Office Suite* and *Instachat 😜* are considered English-speaking, which solves our problem.

Let's now apply this function to the entire dataset in order to remove all non-English-speaking apps. We may accidentally remove some English-speaking apps that have more than 3 emoji in the title, but we should have enough rows of data to work with (as we'll see later).

In [19]:
goog_english = [] # Empty lists that will contain our
app_english = []  # English-speaking apps for both datasets.

for row in android_clean:
    name = row[0]
    if english_or_not(name):
        goog_english.append(row) # If it's English, add it to the list.
        
for row in app_data[1:]: # Exclude the header row from app_data.
    name = row[1]
    if english_or_not(name):
        app_english.append(row)
        
# Now we should have two list-of-lists of only English-speaking apps.
# By printing the length of these, we should see how many apps we have
# left in each dataset. We can also use arithmetic to see how many apps
# we deleted in the process.

print(len(goog_english))
print(len(app_english), "\n")

print(len(android_clean) - len(goog_english))
print(len(app_data[1:]) - len(app_english))

9614
6183 

45
1014


## 6. Isolating Free Apps

Using the same method we did before to isolate English-speaking apps, we'll isolate **free** English-speaking apps. We'll take each app's price and compare it to *0* and *0.0* for the Google and Apple datasets respectively, then add the app entry to the new dataset only if its price is free.

In [20]:
goog_eng_free = []
app_eng_free = []

for row in goog_english:
    price = row[7] # Index 7 is the Price value as an float.
    if price == '0':
        goog_eng_free.append(row)
        
for row in app_english:
    price = row[4] # Index 4 is the price value as a float.
    if price == '0.0':
        app_eng_free.append(row)
        
# Just like before, now we should now have a smaller dataset
# which only contains free, English-speaking apps.
        
print(len(goog_eng_free))
print(len(app_eng_free))

8864
3222


## 7. Building Frequency Tables by Genre
Now that we've cleaned errors and duplicates, removed non-English entries, and removed paid apps from our datasets, we can finally analyze the material and determine what our company's best interest is. They provided us with their strategy so that we can use it to guide our analysis.

1. __Build a minimal Android version of the app, and add it to Google Play.__

2. __If the app has a good response from users, we develop it further.__

3. __If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.__

Since our company's goal envisions an app that succeeds on both the Google and Apple market, we'll need to find apps that succeed in both markets. We can start by finding the most common genres for apps in each dataset by building a frequency table function.

To do this, we'll create an empty dictionary and loop through each dataset using the *prime_genre* value as the **key** for the Apple dataset, the *Category* and *Genres* values as the **keys** for the Google dataset, and the number of instances as the **value pair** for both datasets.

However, dictionaries do not preserve order, so we'll have to build a second function that will properly order our frequency table so we can easily see which genres make up the highest percentage of the market and help our company make a decision. We'll find the percentages by dividing the number of instances of each genre by the total number of apps in each dataset.

In [21]:
def build_a_freq_table(your_dataset, your_index):
    dictionary = {}  # This dictionary is only defined in our function.
    counter = 0      # Initialize a temporary counter variable to zero.
    
    for row in your_dataset:
        counter += 1 # This will count how many rows you have for %'s.
        value = row[your_index]
        
        # This if-else statement will either create a key-value pair
        # or add one to a current value pair depending on if the 
        # current row is already in the dictionary. This way, you
        # get a final dictionary with the total number of iterations
        # of your_index, which in our case is Genres, prime_genre, etc.
        
        if value in dictionary:
            dictionary[value] += 1
        else:
            dictionary[value] = 1
    
    percentage_dict = {} # This dictionary will be basically the same,
                         # but it will have percentages as the value
                         # instead of the number of iterations.
    for key in dictionary:
        percent = (dictionary[key] / counter) * 100
        
        # This arithmetic divides each value of a key by the total
        # number of keys. For example, this will give us the percentage  
        # of gaming apps out of all other genres.
        
        percentage_dict[key] = percent
        
        # This sets the value of each key to a percentage instead
        # of a total number of iterations.
    
    return percentage_dict

Our company provided a *display_table()* function for us that will organize the frequency table by descending percentages.

In [22]:
def display_table(dataset, index):
    table = build_a_freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Let's use our functions to display the genre frequency tables for the Apple and Google datasets. To do this, we'll need to pass in each dataset as the first inputs of the function, and the index of *prime_genre*, *Genres*, and *Category* for the second inputs of the function.

In [23]:
display_table(app_eng_free, 11)
# This one is based on the prime_genre column of the Apple dataset.

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


In [24]:
display_table(goog_eng_free, 1)
# This one is based on the Category column of the Google dataset.

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [25]:
display_table(goog_eng_free, 9)
# This one is based on the Genres column of the Google dataset.
# Recall that there can be apps that have multiple genres.

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

We've successfuly created 3 frequency tables that all relate to the genres of apps on the Apple and Google market. Let's analyze them one by one, starting from the top.

### A. The App Store Market

The Apple market is dominated by *Games* apps with over half of the marketshare at ~60%. *Entertainment* is at a far-away second place percentage of ~8%. 

When you add in the percentage of *Entertainment*, *Photo & Video*, *Social Networking*, *Sports*, and *Music* apps to the ~60% marketshare that *Games* have, the number jumps up to ~80%. So basically, 4 out of every 5 apps on the Apple Store are entertainment based. 

### B. The Google Play Market

Compared to the Apple dataset, this dataset has much more variety of genres. The most common category is *Family* and *Game* apps, but there is a sizable chunk of non-entertainment apps in the marketshare. 

More surprisingly, the most common genre of app is *Tools* at ~8% of the marketshare. Second place is *Entertainment* at ~6%. There are a lot of mixed genres that games have, but it is important to note that if our company decided to make a *Business* or *Productivity* app in the Google market, we would still have a chance to reach people.

### C. Making a Decision

For our company to succeed in the Apple market, we'll need to produce entertainment apps, preferably games, in order to reach the largest number of people for ad revenue.

While the choice to make games and family apps isn't as dire in the Google market, it's still the best choice for our company to reach the most people in the Google market as well.

Since our company's strategy is to **start** in the Google market and then proceed to the Apple market if and only if the app is profitable after 6 months, this gives our company the chance to see if our app can profit in the smaller percentage of market that games have in the Google Play market first before we move to the giant gaming marketshare of the App Store.

Our entertainment apps will have to succeed in the widely varied Google Play market. It's not easy for us to assume that if they succeed there, they will succeed in the saturated Apple Store market as well, but in order to reach the large market of the App Store, we will have to take risks as a company and produce games.

Also, these frequency tables only show how **many** apps of each genre are on the market and don't show us how many users **actually downloaded** these apps. So, we'll need to do some further research to narrow our decision to more specific types of apps in the game area.

## 8. Building Frequency Tables by Most Popular Genre

By looking at the Google dataset documentation, we know that the column at index *5* is for *Installs*, or the number of downloads. But we don't have this information for the Apple dataset. So we'll get around that by using the *rating_count_tot* column at index *5* as a substitute.

### A. Most Popular by Genre in the App Store

First, we'll loop through the genres frequency table that we just created. Each time we do this, we'll set two variables to zero: the total number of user ratings, and the number of apps per genre. 

Then, we'll loop through every entry in the cleaned Apple dataset and see if we find a match. We will be on a certain genre, so if an app matches it, we'll store it into a variable and then iterate a variable by one to count that app. We'll also add the number of ratings that app has to the total.

Once we have the total user ratings and number of apps, we can divide them respectively to get the average number of user ratings per genre. This will give us an idea of which genres are the most popular with users.

In [26]:
# We need to generate a frequency table and store it in
# a dictionary named genres_freq_table. We'll use our 
# function from before to generate it. 

genres_freq_table = build_a_freq_table(app_eng_free, 11)

# Now, we can loop through this frequency table by each genre.
# In another for loop, we see if we can find an app that has
# the same genre as the first for loop, and if we do, we
# update the total and len_genre variable. Eventually, every
# genre from the first for loop will be matched with an app
# from the second for loop.

for genre in genres_freq_table:
    
    total = 0     # This counter will store the sum of user ratings.
    len_genre = 0 # This counter will store the number of apps per genre.
    
    for row in app_eng_free:
        genre_app = row[11] # Index 11 is the prime_genre value.
        
        if genre_app == genre:            
            num_ratings = float(row[5]) # Index 5 is the rating_count_tot value.
            total += num_ratings
            len_genre += 1
            
    average = total / len_genre
    print(genre, '-', average)

Social Networking - 71548.34905660378
Photo & Video - 28441.54375
Games - 22788.6696905016
Music - 57326.530303030304
Reference - 74942.11111111111
Health & Fitness - 23298.015384615384
Weather - 52279.892857142855
Utilities - 18684.456790123455
Travel - 28243.8
Shopping - 26919.690476190477
News - 21248.023255813954
Navigation - 86090.33333333333
Lifestyle - 16485.764705882353
Entertainment - 14029.830708661417
Food & Drink - 33333.92307692308
Sports - 23008.898550724636
Book - 39758.5
Finance - 31467.944444444445
Education - 7003.983050847458
Productivity - 21028.410714285714
Business - 7491.117647058823
Catalogs - 4004.0
Medical - 612.0


From these results, we can see that despite *Games* having the largest marketshare, they do not have the highest number of user ratings. Even though this doesn't really show us how many people downloaded them, we can get an idea that games aren't necessarily the most popular apps.

We can recommend to our company that entertainment apps focused on entertainment besides games would be a better approach, but something to note is that non-entertainment apps have a large amount of ratings as well. 

We can also recommend our company to make *Reference*, *Utility*, *Travel*, *Shopping*, *Lifestyle*, etc. apps and dominate that small portion of the marketplace to get the most ad revenue, but this is contrary to our first hypothesis. The gaming market **is** over-saturated and we might not have the best chance to create the next viral sensation, but abandoning the entertainment area entirely for one of these niche, non-entertainment sections may not be the best either.

Since our company wants to start with the Google Play market first, let's postpone this decision until we look further through the Google Play dataset.

### B. Most Popular by Genre in Google Play

The technique for finding this frequency table is very similar to how we did it for the Apple dataset. The only difference is that the value for *Installs* is stored as a string, and has problems characters such as '+' and ',' mixed in with the numbers we need. We have to remove those and convert the *string* to a *float* so that we can use the value in calculations like before.

Another problem is that besides the string conversion, the download data is very broad and only gives us a range of user downloads. We'll just have to take the number given to us as an estimate of the amount of downloads. We're doing this to get an idea of genre popularity, so accuracy isn't too important, and an estimation will suffice.

We can use either the *Categories* or *Genres* column for this frequency table, but since the *Categories* column doesn't have mixed entries, we'll use that for now.

In [27]:
# Generate the frequency table we'll be working on.
categories_freq_table = build_a_freq_table(goog_eng_free, 1)

# Loop through each category, and just like before, update
# total and len_category when we find a match with row in
# the second for loop.

for category in categories_freq_table:
    
    total = 0
    len_category = 0
    
    for row in goog_eng_free:
        category_app = row[1] # Index 1 is the Category value.
       
        # If we find a match, we need to do some changes to
        # the string (num_installs). First, we need to replace all
        # of the commas and plus signs with '' in order to remove
        # them. Then, everything works just like the last example
        # once we convert the string into a float type.
        
        if category_app == category:            
            
            num_installs = row[5] # Index 5 is the Installs value.
            num_installs = num_installs.replace(',', '')
            num_installs = num_installs.replace('+', '')
            
            total += float(num_installs)
            len_category += 1
            
    average = total / len_category
    print(category, '-', average)

ART_AND_DESIGN - 1986335.0877192982
AUTO_AND_VEHICLES - 647317.8170731707
BEAUTY - 513151.88679245283
BOOKS_AND_REFERENCE - 8767811.894736841
BUSINESS - 1712290.1474201474
COMICS - 817657.2727272727
COMMUNICATION - 38456119.167247385
DATING - 854028.8303030303
EDUCATION - 1833495.145631068
ENTERTAINMENT - 11640705.88235294
EVENTS - 253542.22222222222
FINANCE - 1387692.475609756
FOOD_AND_DRINK - 1924897.7363636363
HEALTH_AND_FITNESS - 4188821.9853479853
HOUSE_AND_HOME - 1331540.5616438356
LIBRARIES_AND_DEMO - 638503.734939759
LIFESTYLE - 1437816.2687861272
GAME - 15588015.603248259
FAMILY - 3695641.8198090694
MEDICAL - 120550.61980830671
SOCIAL - 23253652.127118643
SHOPPING - 7036877.311557789
PHOTOGRAPHY - 17840110.40229885
SPORTS - 3638640.1428571427
TRAVEL_AND_LOCAL - 13984077.710144928
TOOLS - 10801391.298666667
PERSONALIZATION - 5201482.6122448975
PRODUCTIVITY - 16787331.344927534
PARENTING - 542603.6206896552
WEATHER - 5074486.197183099
VIDEO_PLAYERS - 24727872.452830188
NEWS_AND_

## 9. Conclusion

It's clear from the Google Play store that utility based apps do have a large amount of downloads, but our original hypothesis of focusing on game and entertainment apps has some basis too despite the saturated market. Clearly the markets for both types of apps are profitable on Google Play, but our company wants to focus on a certain type of app. 

So what to do? If games and entertainment are based on the rare ability to successfuly make viral content, and utility apps don't always work cross-platform, what do we choose to recommend to our company? *Medical*, *Parenting*, *Weather*, *Sports*, *Lifestyle*, *Health & Fitness*, and *Dating* are all situationally useful. *Most people aren't guaranteed to keep using these apps once downloaded*. We need to recommend something to our company that people will constantly use on the Google Play store that will translate to a large userbase in the App Store.

Since *Navigation* and *Education* apps require a large degree of knowledge to produce, let's narrow it down to the next best option, and it's a lucrative one. If our company chooses to focus in the *Books & Reference* genre, we would reach a large userbase in both markets and succeed in accruing ad revenue on a constant basis. However, the are already industry giants that are inflating the statistics of this genre, so it's best to stay away, despite how good it seems on paper.

So our final decision is to recommend making gaming apps to our company, but we need a little more feedback before we can determine what types of games to make.


## 10. Narrowing Down the Final Recommendation

After pitching the idea of making free games, the company changed it's position. They feel that typical free games don't last as long on the market as they would want them to, so they are open to paid content to keep users coming back. However, they only want to do so if there is enough positive feedback. Therefore, we need to find apps that have high user ratings and high download quantities so that we can narrow down what type of game to make.

We haven't generated a frequency table of the *Genres* column of the Google dataset yet, so we can use this to help us find out. Once we create this frequency table, we'll use the *Ratings* column to determine which genres of games are the most well rated overall in the Google market and recommend some to the company.

In [28]:
# Generate the frequency table we'll be working on.
genres_freq_table = build_a_freq_table(goog_eng_free, 9)

# Loop through each category, and just like before, update
# total and len_genre when we find a match with row in
# the second for loop.

for genre in genres_freq_table:
    
    total = 0
    len_genre = 0
    
    for row in goog_eng_free:
        genre_app = row[9] # Index 9 is the Genre value.
       
        # If we find a match, we need to do some changes to
        # the string (num_installs). First, we need to replace all
        # of the commas and plus signs with '' in order to remove
        # them. Then, everything works just like the last example
        # once we convert the string into a float type.
        
        if genre_app == genre:            
            
            num_installs = row[5] # Index 5 is the Installs value.
            num_installs = num_installs.replace(',', '')
            num_installs = num_installs.replace('+', '')
            
            total += float(num_installs)
            len_genre += 1
            
    average = total / len_genre
    print(genre, '-', average)

Art & Design - 2122850.9433962265
Art & Design;Creativity - 285000.0
Auto & Vehicles - 647317.8170731707
Beauty - 513151.88679245283
Books & Reference - 8767811.894736841
Business - 1712290.1474201474
Comics - 831873.1481481482
Comics;Creativity - 50000.0
Communication - 38456119.167247385
Dating - 854028.8303030303
Education - 550185.4430379746
Education;Creativity - 2875000.0
Education;Education - 4759517.0
Education;Pretend Play - 1800000.0
Education;Brain Games - 5333333.333333333
Entertainment - 5602792.775092937
Entertainment;Brain Games - 3314285.714285714
Entertainment;Creativity - 4000000.0
Entertainment;Music & Video - 6413333.333333333
Events - 253542.22222222222
Finance - 1387692.475609756
Food & Drink - 1924897.7363636363
Health & Fitness - 4188821.9853479853
House & Home - 1331540.5616438356
Libraries & Demo - 638503.734939759
Lifestyle - 1412998.3449275363
Lifestyle;Pretend Play - 10000000.0
Card - 3815462.5
Arcade - 22888365.48780488
Puzzle - 8302861.91
Racing - 1591064

From this frequency table, there's no real observable pattern, like hypothesized, since downloads are spread out evenly over the genres. We'll instead generate a frequency table based off of app ratings and see which genres are the most liked.

In [29]:
# Generate the frequency table we'll be working on.
rate_genres_freq_table = build_a_freq_table(goog_eng_free, 9)

# Now, we can loop through this frequency table by each genre.
# In another for loop, we see if we can find an app that has
# the same genre as the first for loop, and if we do, we
# update the total and len_genre variable. Eventually, every
# genre from the first for loop will be matched with an app
# from the second for loop.

for genre in rate_genres_freq_table:
    
    total = 0     # This counter will store the sum of user ratings.
    len_genre = 0 # This counter will store the number of apps per genre.
    
    for row in goog_eng_free:
        genre_app = row[9] # Index 9 is the Genre value.
        
        if genre_app == genre:
            
            # After glancing at the dataset, there are many values where
            # instead of a rating, it says NaN, or Not a Number. We will
            # need to get rid of these values or it will ruin or analysis.
            
            if row[2] == "NaN": # Index 2 is the Rating value.
                rating_app = 0  # This prevents adding NaN to our total. 

            else:
                rating_app = float(row[2]) # If it isn't NaN, make the rating
                len_genre += 1 # a float and iterate the length counter by 1.
            total += rating_app # Add the rating to the total.
    
    if len_genre != 0: # This checks that we don't divide by zero.
        average = total / len_genre # And this is the average of the total
                                    # number of ratings over how many apps
                                    # are in the genre.
        
    print(genre, '-', average)

Art & Design - 4.3352941176470585
Art & Design;Creativity - 4.35
Auto & Vehicles - 4.184722222222223
Beauty - 4.278571428571428
Books & Reference - 4.347798742138364
Business - 4.10395256916996
Comics - 4.165384615384617
Comics;Creativity - 4.8
Communication - 4.126923076923076
Dating - 3.9824427480916027
Education - 4.294117647058822
Education;Creativity - 4.375
Education;Education - 4.303333333333332
Education;Pretend Play - 4.1
Education;Brain Games - 4.433333333333334
Entertainment - 4.0734065934065935
Entertainment;Brain Games - 4.3
Entertainment;Creativity - 4.533333333333333
Entertainment;Music & Video - 4.180000000000001
Events - 4.435555555555557
Finance - 4.128373702422146
Food & Drink - 4.1673913043478255
Health & Fitness - 4.236051502145922
House & Home - 4.140983606557378
Libraries & Demo - 4.178125
Lifestyle - 4.082374100719423
Lifestyle;Pretend Play - 4.0
Card - 4.031578947368422
Arcade - 4.266233766233768
Puzzle - 4.355421686746987
Racing - 4.147619047619049
Sports - 4.

## 11. Final Recommendation and Conclusion

By observing the ratings, we can see that *Puzzle*, *Strategy*, *Brain*, *etc.* games seem to get the most positive reviews. This narrows are focus and allows us to recommend these types of apps to the company, and with large user bases and highly positive feedback, it would be in our best interest to offer paid content in the form of extra puzzles or challenges to increase revenue.

Therefore, our final recommendation to the company is to produce free *Puzzle*, *Strategy*, *Brain*, *etc.* based games that require watching ads to continue, and/or purchasing extra content in order to remove ads or play exclusive content. If our game becomes profitable after a 6 month period on the Google Play store, it will be able to enjoy a large userbase in the App Store as well. Because these types of games are so well reviewed, we can be confident that users will not complain or lose interest in our apps because of ads or paid content.